modules/rostests/kmtests/ntos_io/IoDeviceInterface.c

```diff
@@ -1,125 +1,214 @@
-      RtlInitUnicodeString(&AliasSymbolicLink, NULL);
-      IoGetDeviceInterfaceAlias(
-         &SymbolicLink,
-         &AliasGuid,
-         &AliasSymbolicLink);
-      if (AliasSymbolicLink.Buffer != NULL)
-      {
-         DPRINT("[PnP Test] Original: %S\n", SymbolicLinkListPtr);
-         DPRINT("[PnP Test] Alias: %S\n", AliasSymbolicLink.Buffer);
-      }
-   }
-#endif
-
-   ExFreePool(SymbolicLinkList);
+    RtlInitUnicodeString(&SymbolicLinkName, SymbolicLink);
+    RtlInitUnicodeString(&KeyName, L"ReactOS_kmtest");
+
+    for (size_t n = 0; n < RTL_NUMBER_OF(Types); ++n)
+    {
+        HANDLE DeviceInterfaceKey, DeviceInterfaceSubKey;
+        OBJECT_ATTRIBUTES ObjectAttributes;
+
+        NTSTATUS Status = IoOpenDeviceInterfaceRegistryKey(&SymbolicLinkName, KEY_CREATE_SUB_KEY, &DeviceInterfaceKey);
+
+        if (NT_SUCCESS(Status))
+        {
+            trace("IoOpenDeviceInterfaceRegistryKey(): success: %d %p\n", n, DeviceInterfaceKey);
+        }
+        else
+        {
+            trace("IoOpenDeviceInterfaceRegistryKey(): fail: %d 0x%x\n", n, Status);
+        }
+
+        InitializeObjectAttributes(&ObjectAttributes,
+                                   &KeyName,
+                                   OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
+                                   DeviceInterfaceKey,
+                                   NULL);
+        Status = ZwCreateKey(&DeviceInterfaceSubKey,
+                             KEY_WRITE,
+                             &ObjectAttributes,
+                             0,
+                             NULL,
+                             REG_OPTION_NON_VOLATILE,
+                             NULL);
+
+        if (NT_SUCCESS(Status))
+        {
+            trace("ZwCreateKey(): successfully created subkey: %d %p\n", n, DeviceInterfaceSubKey);
+
+            ZwDeleteKey(DeviceInterfaceSubKey);
+            ZwClose(DeviceInterfaceSubKey);
+        }
+        else
+        {
+            trace("ZwCreateKey(): failed to create a subkey: %d 0x%x\n", n, Status);
+        }
+
+        ZwClose(DeviceInterfaceKey);
+    }
 }
 
 static
 VOID
-Test_IoRegisterDeviceInterface(VOID)
+Test_IoGetDeviceInterfaceAlias(PWSTR SymbolicLink)
 {
-    GUID Guid = {0x378de44c, 0x56ef, 0x11d1, {0xbc, 0x8c, 0x00, 0xa0, 0xc9, 0x14, 0x05, 0xdd}};
-    DEVICE_OBJECT DeviceObject;
-    EXTENDED_DEVOBJ_EXTENSION DeviceObjectExtension;
-    DEVICE_NODE DeviceNode;
     UNICODE_STRING SymbolicLinkName;
-    NTSTATUS Status;
 
-    RtlInitUnicodeString(&SymbolicLinkName, L"");
+    RtlInitUnicodeString(&SymbolicLinkName, SymbolicLink);
+
+    for (size_t n = 0; n < RTL_NUMBER_OF(Types); ++n)
+    {
+        UNICODE_STRING AliasSymbolicLinkName;
+        NTSTATUS Status = IoGetDeviceInterfaceAlias(&SymbolicLinkName, Types[n], &AliasSymbolicLinkName);
+
+        if (NT_SUCCESS(Status))
+        {
+            trace("IoGetDeviceInterfaceAlias(): success: %d %wZ\n", n, &AliasSymbolicLinkName);
```

Comment:
Like above, you're not testing much here.
Does the call succeed if you pass in the same GUID that was originally passed to `IoGetDeviceInterfaces`? If so, then you can pass the GUID (or the `n`) through to this function, and if they're the same, expect success. And presumably we'd expect the Alias name to be the same string as the original in that case?

If the function call fails, what status do we expect? Presumably `STATUS_OBJECT_NAME_NOT_FOUND` or `STATUS_OBJECT_PATH_NOT_FOUND`?
If you pass in a custom GUID as the second parameter, do you get `STATUS_INVALID_HANDLE` like the documentation seems to suggest?

These are all things you can codify into `ok()`s to make the test actually verify something, instead of just tracing.

Reply:
>Does the call succeed if you pass in the same GUID that was originally passed to IoGetDeviceInterfaces? If so, then you can pass the GUID (or the n) through to this function, and if they're the same, expect success. And presumably we'd expect the Alias name to be the same string as the original in that case?

As you might notice, I'm always passing the same list of GUIDs into all functions, independently from `IoGetDeviceInterfaces`. So they should use those GUIDs separately.
What I'm only passing from `IoGetDeviceInterfaces`, is actually **symbolic link**, not a GUID, since MSDN says that the symbolic links should be obtained from `IoGetDeviceInterfaces` or `IoRegisterDeviceInterface` first.
However, I've also added a case which passes alias symbolic links into `IoOpenDeviceInterfaceRegistryKey`, since the documentation says that the symbolic link for that function can be also obtained from `IoGetDeviceInterfaceAlias`. And additionally, I added a NULL GUID to the list of GUIDs, which can be considered as invalid or, as you said, custom. Or custom GUID should be only declared by myself and shouldn't exist anywhere else? :thinking: 
See an updated code for details. :slightly_smiling_face:

Reply:
>If the function call fails, what status do we expect? Presumably STATUS_OBJECT_NAME_NOT_FOUND or STATUS_OBJECT_PATH_NOT_FOUND?

At least when I'm passing `GUID_NULL`, it fails with 0xc0000034 (`STATUS_OBJECT_NAME_NOT_FOUND`), obviously because it was not found in registry. But the GUID itself seems to be valid (as parameter).
![test_results](https://user-images.githubusercontent.com/26385117/149595010-7e321845-b780-42d1-a72d-b57e59b4f0fd.png)
However I didn't check yet what happens if one of the parameters (either symbolic link or GUID) is really invalid (`NULL`).

Reply:
OK, just added an invalid parameter testcases.
At least when I'm passing an invalid (`NULL`) symbolic link, it returns `STATUS_INVALID_PARAMETER`, not `STATUS_INVALID_HANDLE`. With invalid GUID the same value should be returned I guess, because MSDN says that it returns identical value if one of both parameters is invalid. So I should return the same in my implementation. :slightly_smiling_face: 
![invalid_parameter](https://user-images.githubusercontent.com/26385117/149626088-4518d4ef-31bb-4a9c-83f9-e7a8eb9ce916.png)
But I still don't understand why MSDN lies about the return value itself. Maybe it has been changed if the future Windows versions, but why then there is no any note about that then?... :thinking: Although these are related return codes, but nevertheless...